add: DevTest pattern for container (#500)
microsoft/code-with-engineering-playbook: automated-testing/container-testing/azdo-container-dev-test-release.md

> # Building Containers with Azure DevOps using DevTest Pattern

Reviewer: Thanks for sharing your learnings. I like this, I think we should add a new section for case studies/examples and place this there.

> In this documents, we highlight learnings from applying the DevTest pattern to container development in Azure DevOps through pipelines.

Reviewer: Can you shed some light on why did you use this vs something else? I think it will be beneficial to understand your reasoning and the advantages of using this approach.

> In this documents, we highlight the learnings from the teams last project regarding how to apply the DevTest pattern to container development in Azure DevOps through pipelines.

Reviewer 2: ```suggestion
In this documents, we highlight learnings from applying the DevTest pattern to container development in Azure DevOps through pipelines.
```

Reviewer 2: slight change suggested - while this "we", and "our last project", might have been relevant in a presentation or internal sharing doc, it sounds a bit weird here - after this PR is merged, who is we? - anytime we use the term we it should refer to all of CSE

Author: You are right, and I was not aware of that. thx.

The pattern enabled as to build container for development, testing and releasing the container for further reuse (production ready).

We will dive into tools needed to build, test and push a container, our environment and go through each step separately.

Follow this link to dive deeper or revisit the [DevTest pattern](https://docs.microsoft.com/en-us/azure/architecture/solution-ideas/articles/dev-test-paas).

## Table of Contents

[Build the Container](#Build-The-Container)
[Test the Container](#Test-the-Container)
[Push Container](#Push-container)
[References](#References)

## Build the Container

> The first step in container development, after creating the necessary Dockerfiles and source code, is building the container. Even the Dockerfile itself can include some basic testing. In our case, the code tests are performed when pushing the code to the repository origin, where it is then used to build the container.

Reviewer 2: ```suggestion
The first step in container development, after creating the necessary Dockerfiles and source code, is building the container. Even the Dockerfile itself can include some basic testing. Code tests are performed when pushing the code to the repository origin, where it is then used to build the container.
```

The first step in our pipeline is to run the `docker build` command with a temporary tag and the required build arguments:

```yaml
- task: Bash@3
  name: BuildImage
  displayName: 'Build the image via docker'
  inputs:
    workingDirectory: "$(System.DefaultWorkingDirectory)${{ parameters.buildDirectory }}"
    targetType: 'inline'
    script: |
      docker build -t ${{ parameters.imageName }} --build-arg YOUR_BUILD_ARG -f ${{ parameters.dockerfileName }} .
  env:
    PredefinedPassword: $(Password)
    NewVariable: "newVariableValue"
```

This task includes the parameters `buildDirectory`, `imageName` and `dockerfileName`, which have to be set beforehand.
This task can for example be used in a template for multiple containers to improve code reuse.

It is also possible to pass environment variables directly to the Dockerfile through the `env` section of the task.

If this task succeeds, the Dockerfile was build without errors and we can continue to testing the container itself.

## Test the Container

> To test the container, we are using the tox environment.

Reviewer: Can you add some context on tox and how it fits in this doc, the link can be left there for those that want to dig deeper.

For more details on tox please visit the tox section of this repository or visit [the official tox documentation page](https://tox.readthedocs.io/en/latest/example/pytest.html).

> Before we test the container though, we are checking for exposed credentials in the docker image history.

Reviewer 2: ```suggestion
Before we test the container, we are checking for exposed credentials in the docker image history.
```

> If our known passwords that we used to access our internal resources, are exposed here, the build step will fail:

Reviewer 2: ```suggestion
If known passwords, used to access our internal resources, are exposed here, the build step will fail:
```

```yml
- task: Bash@3
  name: CheckIfPasswordInDockerHistory
  displayName: 'Check for password in docker history'
  inputs:
    workingDirectory: "$(System.DefaultWorkingDirectory)"
    targetType: 'inline'
    failOnStdErr: true
    script: |
      if docker image history --no-trunc ${{ parameters.imageName }} | grep -qF $PredefinedPassword; then
        exit 1;
      fi
      exit 0;
  env:
    PredefinedPassword: $(Password)
```

> After our credential test, the container is tested through the pytest extension [testinfra](https://testinfra.readthedocs.io/en/latest/).

Reviewer 2: ```suggestion
After the credential test, the container is tested through the pytest extension [testinfra](https://testinfra.readthedocs.io/en/latest/).
```

# Building Containers with Azure DevOps using DevTest Pattern

> In this documents, we highlight the learnings from the teams last project regarding how to apply
> the DevTest pattern to container development in Azure DevOps through pipelines.

Reviewer 3: ```suggestion
In this documents, we highlight the learnings from the teams last project regarding how to apply the DevTest pattern to container development in Azure DevOps through pipelines.
```

> The pattern enabled as to build container for development, testing
> and releasing the container for further reuse (production ready).

Reviewer 3: ```suggestion
The pattern enabled as to build container for development, testing and releasing the container for further reuse (production ready).
```

We will dive into tools needed to build, test and push a container, our environment and go through each step separately.

Follow this link to dive deeper or revisit the [DevTest pattern](https://docs.microsoft.com/en-us/azure/architecture/solution-ideas/articles/dev-test-paas).

## Table of Contents

[Build the Container](#Build-The-Container)
[Test the Container](#Test-the-Container)
[Push Container](#Push-container)
[References](#References)

## Build the Container

> The first step in container development, after creating the necessary Dockerfiles and source code, is building the container.
> Even the Dockerfile itself can include some basic testing.
> In our case, the code tests are performed when pushing the code to the repository origin,
> where it is then used to build the container.

Reviewer 3: ```suggestion
The first step in container development, after creating the necessary Dockerfiles and source code, is building the container. Even the Dockerfile itself can include some basic testing. In our case, the code tests are performed when pushing the code to the repository origin, where it is then used to build the container.
```

If this task succeeds, the Dockerfile was build without errors and we can continue to testing the container itself.

## Test the Container

To test the container, we are using the tox environment.
For more details on tox please visit the tox section of this repository or visit [the official tox documentation page](https://tox.readthedocs.io/en/latest/example/pytest.html).

Before we test the container though, we are checking for exposed credentials in the docker image history.
If our known passwords that we used to access our internal resources, are exposed here, the build step will fail:

```yml
- task: Bash@3
  name: CheckIfPasswordInDockerHistory
  displayName: 'Check for password in docker history'
  inputs:
    workingDirectory: "$(System.DefaultWorkingDirectory)"
    targetType: 'inline'
    failOnStdErr: true
    script: |
      if docker image history --no-trunc ${{ parameters.imageName }} | grep -qF $PredefinedPassword; then
        exit 1;
      fi
      exit 0;
  env:
    PredefinedPassword: $(Password)
```

After our credential test, the container is tested through the pytest extension [testinfra](https://testinfra.readthedocs.io/en/latest/).

> Testinfra is a Python-based tool which can be used to start a container, gather prerequisites,
> test the container and shut it down again, without any effort besides writing the tests.
> These tests can for example include:

Reviewer 3: ```suggestion
Testinfra is a Python-based tool which can be used to start a container, gather prerequisites, test the container and shut it down again, without any effort besides writing the tests. These tests can for example include:
```

    assert release == host.system_info.release

def extract_env_var(file_content):
    import re

    regex = r"ENV_VAR=\"(?P<s>[^\"]*)\""

    match = re.match(regex, file_content)
    return match.group('s')

def test_ports_exposed(host):
    port1 = "9010"
    st1 = f"grep -q {port1} /app/Dockerfile && echo 'true' || echo 'false'"
    cmd1 = host.run(st1)
    assert cmd1.stdout

def test_listening_simserver_sockets(host):
    assert host.socket("tcp://0.0.0.0:32512").is_listening
    assert host.socket("tcp://0.0.0.0:32513").is_listening
```

To start the test, a [pytest](http://pytest.org) command is executed through tox.

A task containing the tox command can look like this:

```yaml
- task: Bash@3
  name: RunTestCommands
  displayName: "Test - Run test commands"
  inputs:
    workingDirectory: "$(System.DefaultWorkingDirectory)"
    targetType: 'inline'
    script: |
      tox -e testinfra-${{ parameters.makeTarget }} -- ${{ parameters.imageName }}
    failOnStderr: true
```

Which could trigger the following pytest code, which is contained in the tox.ini file:

```bash
pytest -vv tests/{env:CONTEXT} --container-image={posargs:{env:IMAGE_TAG}} --volume={env:VOLUME}
```

> As a last task of this pipeline to build and test the container,
> we set a variable called `testsPassed` which is only `true`, if the previous tasks succeeded:

Reviewer 3: ```suggestion
As a last task of this pipeline to build and test the container, we set a variable called `testsPassed` which is only `true`, if the previous tasks succeeded:
```

```yml
- task: Bash@3
  name: UpdateTestResultVariable
  condition: succeeded()
  inputs:
    targetType: 'inline'
    script: |
      echo '##vso[task.setvariable variable=testsPassed]true'
```

## Push container

> After building and testing, if our container runs as expected, we want to release it to our
> Azure Container Registry (ACR) to be used by our larger application.
> Before that, we want to automate the push behavior and define a meaningful tag.

Reviewer 3: ```suggestion
After building and testing, if our container runs as expected, we want to release it to our Azure Container Registry (ACR) to be used by our larger application. Before that, we want to automate the push behavior and define a meaningful tag.
```
